tests/test_websocket.py

```diff
@@ -0,0 +1,348 @@
+
+import pytest
+from lomond.errors import ProtocolError, HandshakeError
+from lomond.events import Binary, Closed, Ping, Pong, Ready, Text
+from lomond.message import Close
+from lomond.opcode import Opcode
+from lomond.response import Response
+from lomond.session import WebsocketSession
+from lomond.stream import WebsocketStream
+from lomond.websocket import WebSocket
+
+
+class FakeSession(object):
+    def __init__(self, *args, **kwargs):
+        self.socket_buffer = []
+
+    def run(self, *args, **kwargs):
+        pass
+
+    def send(self, opcode, bytes):
+        self.socket_buffer.append((opcode, bytes))
+
+
+@pytest.fixture
+def websocket(monkeypatch):
+    monkeypatch.setattr(
+        'os.urandom', lambda len: b'\x00' * len)
+
+    ws = WebSocket('ws://example.com')
+    return ws
+
+
+@pytest.fixture
+def websocket_with_fake_session(monkeypatch):
+    monkeypatch.setattr(
+        'os.urandom', lambda len: b'\x00' * len)
+
+    ws = WebSocket('ws://example.com')
+    ws.state.session = FakeSession()
+    return ws
+
+
+def generate_data(*frames):
+    payload = [
+        b'HTTP/1.1 101 Switching Protocols\r\n',
+        b'Upgrade: websocket\r\n',
+        b'Connection: Upgrade\r\n',
+        b'User-Agent: Test\r\n',
+        b'Sec-WebSocket-Key: AAAAAAAAAAAAAAAAAAAAAA==\r\n',
+        b'Sec-WebSocket-Accept: icx+yqv66kxgm0fcwalwlflwtai=\r\n',
+        b'\r\n'
+    ]
+    payload.extend(frames)
+
+    return b''.join(payload)
+
+
+def test_init(websocket):
+    assert isinstance(websocket, WebSocket)
+    assert isinstance(websocket.state, WebSocket.State)
+    assert websocket.resource == '/'
+    assert len(b64decode(websocket.key)) == 16
+    assert websocket.session is None
+    assert isinstance(websocket.stream, WebsocketStream)
+
+
+def test_init_with_query():
+    ws = WebSocket('ws://example.com/resource?query')
+    assert ws.resource == '/resource?query'
+
+
+def test_repr(websocket):
+    assert repr(websocket) == "WebSocket('ws://example.com')"
+
+
+def test_port_has_to_be_numeric():
+    with pytest.raises(ValueError) as e:
+        WebSocket('ws://example.com:abc')
+
+    assert str(e.value) == 'illegal port value'
+
+
+def test_is_secure(websocket):
+    assert websocket.is_secure is False
+    assert WebSocket('wss://example.com').is_secure is True
+
+
+def test_get_request(websocket):
+    assert websocket.get_request() == (
+        b'GET / HTTP/1.1\r\n'
+        b'Host: example.com:80\r\n'
+        b'Upgrade: websocket\r\n'
+        b'Connection: Upgrade\r\n'
+        b'Sec-WebSocket-Protocol: \r\n'
+        b'Sec-WebSocket-Key: AAAAAAAAAAAAAAAAAAAAAA==\r\n'
+        #                    ^^^^^^^^^^^^^^^^^^^^^^^^
+        #                     b64encode('\x00' * 16)
+        b'Sec-WebSocket-Version: 13\r\n'
+        b'User-Agent: DataplicityLomond/0.1\r\n'
+        b'\r\n'
+    )
+
+
+def test_connect(websocket):
+    websocket.connect()
+    assert isinstance(websocket.session, WebsocketSession)
```

Comment:
ok, I'll modify the test, but I'd say it's a bit slippery. What I mean by that is that `self.reset()` is the very first thing which happens in connect(), there is no if statement or anything. Also, the default constructor of State does not set the attribute of `session` - only the `connect` call does so. And if run() would not be invoked, I assume we would have some form of exception?